### ISSUE TITLE
run loop not initialized with napari.Viewer()

### ISSUE THREAD

**Issue by DannyAlas** [napari/napari#5382]
## 🐛 Bug

When initializing napari inside a qt application using `napari.Viewer()` it appears as though the run loop for vispy isn't initiated. Attempting to drag and drop an image into the viewer directly after opening the application results in an error, see error below. 

This issue does not occur if `napari.run()` is used after the viewer is initialized, but this seems to start a completely separate instance of napari. Any manipulation of the viewer (for example: adding a dock widget) does not affect the visible napari instance if `napari.run()` is used.

## To Reproduce

1. start napari in qt application using `napari.Viewer()`
2. drag and drop an image into the viewer
```
import napari
import sys
from PyQt5.QtWidgets import QMainWindow, QVBoxLayout, QWidget, QApplication

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        # start qt app
        self._main = QWidget()
        self.setCentralWidget(self._main)
        self.layout = QVBoxLayout(self._main)

        # start napari
        self.viewer = napari.Viewer()

        # add napari to qt apps layout
        napari_qt_window = self.viewer.window._qt_window
        self.layout.addWidget(napari_qt_window)

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = MainWindow()
    window.show()
    app.exec_()
```

## Error

```
30-Nov-22 13:08:44 - vispy    - WARNING  - Traceback (most recent call last):
  File "C:\dev\temp_venv\lib\site-packages\napari\_qt\widgets\qt_welcome.py", line 139, in dragEnterEvent     
    viewer = self.parentWidget().nativeParentWidget()._qt_viewer
AttributeError: 'MainWindow' object has no attribute '_qt_viewer'
```

## Expected behavior

Ideally `napari.Viewer()` should open the viewer and begin the run loop.

## Environment

napari: 0.4.17
Platform: Windows-10-10.0.22621-SP0
Python: 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.7
NumPy: 1.23.5
SciPy: 1.9.3
Dask: 2022.11.1
VisPy: 0.11.0
magicgui: 0.6.1
superqt: 0.4.0
in-n-out: 0.1.6
app-model: 0.1.1
npe2: 0.6.1

OpenGL:
- GL version: 4.6.0 NVIDIA 526.98
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0
- screen 2: resolution 2560x1440, scale 1.0

**Comment by Czaki** [comment]
I see multiple problems here. 

As I see in this code, you would like only Viewer without the rest of GUI.

In such a situation, you should not use `napari.Viewer` but `napari.components.ViewerModel` with `napari.qt.QtViewer`.


```python
from napari.components import ViewerModel
from napari.qt import QtViewer 

viewer_model = ViewerModel()
qt_viewer = QtViewer(viewer_model)
```

The second thing is that viewer cannot activate a loop, as loop activation means stopping the execution of code in the current context and start waiting on incoming events. This is why there are requirements to call `napari.run()` or `app.exec()`




> ## Error
> 
> ```
> 30-Nov-22 13:08:44 - vispy    - WARNING  - Traceback (most recent call last):
>   File "C:\dev\temp_venv\lib\site-packages\napari\_qt\widgets\qt_welcome.py", line 139, in dragEnterEvent     
>     viewer = self.parentWidget().nativeParentWidget()._qt_viewer
> AttributeError: 'MainWindow' object has no attribute '_qt_viewer'
> ```

This error only means that your MainWindow class instance does not have the `_qt_viewer` attribute.

**Comment by DannyAlas** [comment]
Thank you so much for the help!

The code you have adds only the Viewer but for our application we would actually like to embed the entire napari application as a QWidget, as mentioned [here](https://forum.image.sc/t/how-to-build-a-pyqt5-app-that-uses-napari-as-a-module-and-pops-up-a-viewer-that-is-controlled-by-the-app/53021). If `napari.Viewer` isn't the way, how could we go about doing this? Apologies for not being clear in my example, I misunderstood the issue.

**Comment by Czaki** [comment]
Could you provide more details? In the linked post, I do not see the requirement to embed the whole app, only the 3D viewer. And `QtViewer` is enough for this (how it is done in PartSeg https://github.com/4DNucleome/PartSeg/). 
Or do you require some more napari widgets?

**Comment by DannyAlas** [comment]
Our initial idea was to implement something similar to how bioimageit implemented [napari](https://github.com/bioimageit/bioimageit_viewer/blob/main/bioimageit_viewer/viewer/napari.py) using `napari.Viewer`. 

We have a larger qt application for multi-dimensional scientific image processing, and as part of the manual adjustment of images, we'd like to integrate the amazing things napari can do. Specifically, manual adjustment of labels and ROI's. Ideally, napari would open with the viewer, layer controls, layer list, etc. (full application) within the QMainWindow of our application with the data provided by our application. Afterward, napari could export the modified data back into our application for further processing. 

I see in #4650 you mention embedding a QMainWindow within a QMainWindow is not recommended, is there another way to go about this?

**Comment by Czaki** [comment]
All basic widgets are available as QtViewer properties:

https://github.com/napari/napari/blob/5ca30ea2a4ef73fb2a2a1da6a93430ca286d1e88/napari/_qt/qt_viewer.py#L303-L327

So you could embed all required controls in your application by docking these few widgets. 

If it is not enough, then I may need more details.

**Comment by DannyAlas** [comment]
Thank you for the help! We went with your proposed solution of docking the widgets in our application.

For the original issue of 
`AttributeError: 'MainWindow' object has no attribute '_qt_viewer'`
All that needed to be done was to initialize the `_qt_viewer` within our app as such:
```
self.viewer = napari.Viewer(show=False)
self._qt_viewer = self.viewer.window._qt_viewer
```
Using the `ViewerModel` wasn't sufficient for our needs as we wanted all the Napari widgets.

Also, [PartSeg](https://github.com/4DNucleome/PartSeg/) is a great example of implementing Napari in an application, we learned much by reviewing the code. Thank you for the recommendation!
